File: cmd/knowledge-service/internal/infrastructure/storage/minio_client.go
package storage

import (
	"context"
	"fmt"
	"io"
	"time"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
)

// MinIOClient MinIO客户端封装
type MinIOClient struct {
	client     *minio.Client
	bucketName string
}

// MinIOConfig MinIO配置
type MinIOConfig struct {
	Endpoint        string
	AccessKeyID     string
	SecretAccessKey string
	BucketName      string
	UseSSL          bool
}

// NewMinIOClient 创建MinIO客户端
func NewMinIOClient(config MinIOConfig) (*MinIOClient, error) {
	// 初始化MinIO客户端
	minioClient, err := minio.New(config.Endpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(config.AccessKeyID, config.SecretAccessKey, ""),
		Secure: config.UseSSL,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to create minio client: %w", err)
	}

	client := &MinIOClient{
		client:     minioClient,
		bucketName: config.BucketName,
	}

	// 确保bucket存在
	if err := client.ensureBucket(context.Background()); err != nil {
		return nil, fmt.Errorf("failed to ensure bucket: %w", err)
	}

	return client, nil
}

// ensureBucket 确保bucket存在
func (c *MinIOClient) ensureBucket(ctx context.Context) error {
	exists, err := c.client.BucketExists(ctx, c.bucketName)
	if err != nil {
		return fmt.Errorf("failed to check bucket existence: %w", err)
	}

	if !exists {
		err = c.client.MakeBucket(ctx, c.bucketName, minio.MakeBucketOptions{})
		if err != nil {
			return fmt.Errorf("failed to create bucket: %w", err)
		}
	}

	return nil
}

// UploadFile 上传文件
func (c *MinIOClient) UploadFile(ctx context.Context, objectName string, reader io.Reader, size int64, contentType string) error {
	_, err := c.client.PutObject(ctx, c.bucketName, objectName, reader, size, minio.PutObjectOptions{
		ContentType: contentType,
		UserMetadata: map[string]string{
			"uploaded-at": time.Now().UTC().Format(time.RFC3339),
		},
	})
	if err != nil {
		return fmt.Errorf("failed to upload file: %w", err)
	}

	return nil
}

// DownloadFile 下载文件
func (c *MinIOClient) DownloadFile(ctx context.Context, objectName string) (io.ReadCloser, error) {
	object, err := c.client.GetObject(ctx, c.bucketName, objectName, minio.GetObjectOptions{})
	if err != nil {
		return nil, fmt.Errorf("failed to download file: %w", err)
	}

	return object, nil
}

// DeleteFile 删除文件
func (c *MinIOClient) DeleteFile(ctx context.Context, objectName string) error {
	err := c.client.RemoveObject(ctx, c.bucketName, objectName, minio.RemoveObjectOptions{})
	if err != nil {
		return fmt.Errorf("failed to delete file: %w", err)
	}

	return nil
}

// GetPresignedURL 获取预签名URL (用于直接下载)
func (c *MinIOClient) GetPresignedURL(ctx context.Context, objectName string, expires time.Duration) (string, error) {
	presignedURL, err := c.client.PresignedGetObject(ctx, c.bucketName, objectName, expires, nil)
	if err != nil {
		return "", fmt.Errorf("failed to generate presigned url: %w", err)
	}

	return presignedURL.String(), nil
}

// GetFileInfo 获取文件信息
func (c *MinIOClient) GetFileInfo(ctx context.Context, objectName string) (*FileInfo, error) {
	stat, err := c.client.StatObject(ctx, c.bucketName, objectName, minio.StatObjectOptions{})
	if err != nil {
		return nil, fmt.Errorf("failed to get file info: %w", err)
	}

	return &FileInfo{
		Name:         stat.Key,
		Size:         stat.Size,
		ContentType:  stat.ContentType,
		LastModified: stat.LastModified,
		ETag:         stat.ETag,
	}, nil
}

// ListFiles 列出文件
func (c *MinIOClient) ListFiles(ctx context.Context, prefix string) ([]*FileInfo, error) {
	var files []*FileInfo

	objectCh := c.client.ListObjects(ctx, c.bucketName, minio.ListObjectsOptions{
		Prefix:    prefix,
		Recursive: true,
	})

	for object := range objectCh {
		if object.Err != nil {
			return nil, fmt.Errorf("failed to list files: %w", object.Err)
		}

		files = append(files, &FileInfo{
			Name:         object.Key,
			Size:         object.Size,
			ContentType:  object.ContentType,
			LastModified: object.LastModified,
			ETag:         object.ETag,
		})
	}

	return files, nil
}

// FileInfo 文件信息
type FileInfo struct {
	Name         string
	Size         int64
	ContentType  string
	LastModified time.Time
	ETag         string
}

File: algo/knowledge-service/app/storage/minio_client.py
"""
MinIO Storage Client

MinIO 对象存储客户端，用于文档文件的存储和检索
"""

import io
import logging
from datetime import timedelta
from typing import Optional

from minio import Minio
from minio.error import S3Error

logger = logging.getLogger(__name__)


class FileInfo:
    """文件信息"""

    def __init__(
        self,
        name: str,
        size: int,
        content_type: str,
        last_modified: str,
        etag: str
    ):
        self.name = name
        self.size = size
        self.content_type = content_type
        self.last_modified = last_modified
        self.etag = etag


class MinIOClient:
    """MinIO 客户端封装"""

    def __init__(
        self,
        endpoint: str,
        access_key: str,
        secret_key: str,
        bucket_name: str,
        secure: bool = False
    ):
        """初始化 MinIO 客户端

        Args:
            endpoint: MinIO 服务地址 (不包含 http://)
            access_key: Access Key
            secret_key: Secret Key
            bucket_name: 存储桶名称
            secure: 是否使用 HTTPS
        """
        self.client = Minio(
            endpoint,
            access_key=access_key,
            secret_key=secret_key,
            secure=secure
        )
        self.bucket_name = bucket_name
        self._ensure_bucket()

    def _ensure_bucket(self) -> None:
        """确保存储桶存在"""
        try:
            if not self.client.bucket_exists(self.bucket_name):
                self.client.make_bucket(self.bucket_name)
                logger.info(f"Created bucket: {self.bucket_name}")
            else:
                logger.info(f"Bucket exists: {self.bucket_name}")
        except S3Error as e:
            logger.error(f"Failed to ensure bucket: {e}")
            raise

    async def upload_file(
        self,
        object_name: str,
        data: bytes,
        content_type: str = "application/octet-stream"
    ) -> None:
        """上传文件

        Args:
            object_name: 对象名称（路径）
            data: 文件数据
            content_type: 内容类型
        """
        try:
            data_stream = io.BytesIO(data)
            self.client.put_object(
                self.bucket_name,
                object_name,
                data_stream,
                length=len(data),
                content_type=content_type
            )
            logger.info(f"Uploaded file: {object_name}")
        except S3Error as e:
            logger.error(f"Failed to upload file {object_name}: {e}")
            raise

    async def download_file(self, object_name: str) -> bytes:
        """下载文件

        Args:
            object_name: 对象名称（路径）

        Returns:
            文件数据
        """
        try:
            response = self.client.get_object(self.bucket_name, object_name)
            data = response.read()
            response.close()
            response.release_conn()
            logger.info(f"Downloaded file: {object_name}")
            return data
        except S3Error as e:
            logger.error(f"Failed to download file {object_name}: {e}")
            raise

    async def delete_file(self, object_name: str) -> None:
        """删除文件

        Args:
            object_name: 对象名称（路径）
        """
        try:
            self.client.remove_object(self.bucket_name, object_name)
            logger.info(f"Deleted file: {object_name}")
        except S3Error as e:
            logger.error(f"Failed to delete file {object_name}: {e}")
            raise

    async def get_presigned_url(
        self,
        object_name: str,
        expires: timedelta = timedelta(hours=1)
    ) -> str:
        """获取预签名 URL

        Args:
            object_name: 对象名称（路径）
            expires: 过期时间

        Returns:
            预签名 URL
        """
        try:
            url = self.client.presigned_get_object(
                self.bucket_name,
                object_name,
                expires=expires
            )
            logger.info(f"Generated presigned URL for: {object_name}")
            return url
        except S3Error as e:
            logger.error(f"Failed to generate presigned URL for {object_name}: {e}")
            raise

    async def get_file_info(self, object_name: str) -> Optional[FileInfo]:
        """获取文件信息

        Args:
            object_name: 对象名称（路径）

        Returns:
            文件信息
        """
        try:
            stat = self.client.stat_object(self.bucket_name, object_name)
            return FileInfo(
                name=stat.object_name,
                size=stat.size,
                content_type=stat.content_type,
                last_modified=stat.last_modified.isoformat(),
                etag=stat.etag
            )
        except S3Error as e:
            logger.error(f"Failed to get file info for {object_name}: {e}")
            return None

    async def list_files(self, prefix: str = "") -> list[FileInfo]:
        """列出文件

        Args:
            prefix: 对象名称前缀

        Returns:
            文件信息列表
        """
        try:
            objects = self.client.list_objects(
                self.bucket_name,
                prefix=prefix,
                recursive=True
            )
            files = []
            for obj in objects:
                files.append(FileInfo(
                    name=obj.object_name,
                    size=obj.size,
                    content_type=obj.content_type or "application/octet-stream",
                    last_modified=obj.last_modified.isoformat(),
                    etag=obj.etag
                ))
            return files
        except S3Error as e:
            logger.error(f"Failed to list files with prefix {prefix}: {e}")
            raise


# 全局单例
_minio_client: Optional[MinIOClient] = None


def get_minio_client(
    endpoint: Optional[str] = None,
    access_key: Optional[str] = None,
    secret_key: Optional[str] = None,
    bucket_name: Optional[str] = None,
    secure: bool = False
) -> MinIOClient:
    """获取 MinIO 客户端单例

    Args:
        endpoint: MinIO 服务地址
        access_key: Access Key
        secret_key: Secret Key
        bucket_name: 存储桶名称
        secure: 是否使用 HTTPS

    Returns:
        MinIO 客户端实例
    """
    global _minio_client

    if _minio_client is None:
        if not all([endpoint, access_key, secret_key, bucket_name]):
            raise ValueError("MinIO client not initialized. Provide all required parameters.")
        _minio_client = MinIOClient(endpoint, access_key, secret_key, bucket_name, secure)

    return _minio_client


async def close_minio_client() -> None:
    """关闭 MinIO 客户端"""
    global _minio_client
    _minio_client = None
    logger.info("MinIO client closed")
